src/common/utils/types_test.cc

```diff
@@ -0,0 +1,49 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#include "src/common/utils/types.h"
+
+#include <iostream>
+#include <string>
+
+#include "src/common/testing/gtest.h"
+
+namespace raksha {
+namespace {
+
+TEST(TypesTest, SelectTypeTrue) {
+  SelectType<true, int, std::string>::Result a = 3;
+  EXPECT_EQ(a, 3);
+}
+
+TEST(TypesTest, SelectTypeFalse) {
+  SelectType<false, int, std::string>::Result a = "Hi";
+  EXPECT_EQ(a, "Hi");
+}
+
+TEST(TypesTest, CopyConstTypeTrue) {
+  CopyConst<true, int> a = 3;
+  // a++; // Should not compile.
```

Comment:
I'd love to find a way to statically assert this but it's not something I've worked out how to do.
Open to ideas.

Reply:
Here's how you'd do it:

```C++
ASSERT_EQ(typeid(&a), typeid(const int *));
ASSERT_NE(typeid(&a), typeid(int *));
```

See https://en.cppreference.com/w/cpp/language/typeid . I'm taking the address of `a` because `typeid` strips the top-level const (and volatile) qualifiers, but it leaves alone any qualifiers below the top level. But because of this quirk, I like to also check that the result is not equal to the non-const qualified thing, just to make sure I haven't messed up somehow.